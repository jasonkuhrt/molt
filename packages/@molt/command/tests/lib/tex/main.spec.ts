import { Tex } from '../../../src/lib/Tex/index.js'
import { describe, expect, it } from 'vitest'

describe(`text`, () => {
  $(`can render text`, Tex.Tex().text(`foo`))
})

describe(`block`, () => {
  describe(`width`, () => {
    describe(`%`, () => {
      $(
        `100%`,
        Tex.Tex({ maxWidth: 100 })
          .block(`foo bar`)
          .block({ border: { top: `-` }, width: `100%` }, `foo`)
      )
    })
  })
  $(`can render a line`, Tex.Tex().block(`foo`))
  describe(`border`, () => {
    $(`top`, Tex.Tex().block({ border: { top: `-` } }, `foo`))
    $(
      `right`,
      Tex.Tex()
        .block({ border: { right: `|` } }, `foo`)
        .block(($) =>
          $.set({ border: { right: `|` } })
            .block(`alpha`)
            .block(`bravo bravo`)
            .block(`charlie charlie charlie`)
        )
    )
    $(`bottom`, Tex.Tex().block({ border: { bottom: `-` } }, `foo`))
    $(`left`, Tex.Tex().block({ border: { left: `|` } }, `foo`))
  })

  $(`can render nothing via null`, Tex.Tex().block(null))

  $(`can render two lines`, Tex.Tex().block(`foo`).block(`bar`))
  describe(`builder`, () => {
    $(
      `can render a line`,
      Tex.Tex().block(($) => $.text(`foo`))
    )
    $(
      `can render nothing via null`,
      Tex.Tex().block(() => null)
    )
  })
  describe(`padding`, () => {
    $(`top`, Tex.Tex().block({ padding: { top: 2 } }, `foo`))
    $(`bottom`, Tex.Tex().block({ padding: { bottom: 2 } }, `foo`))
    $(
      `left`,
      Tex.Tex().block(($) => $.set({ padding: { left: 2 } }).text(`foo`))
    )
    $(
      `left multi line`,
      Tex.Tex().block(($) =>
        $.set({ padding: { left: 2 } })
          .block(`foo`)
          .block(`bar`)
          .block(`qux`)
      )
    )
  })
})

describe(`list`, () => {
  $(`can render items`, Tex.Tex().list([`foo`, `bar`]))
  $(`can render multi-line items`, Tex.Tex().list([`foo`, `bar\nbaz\nqux`, `zod`]))
  $(`can have custom bullet`, Tex.Tex().list({ bullet: { graphic: `-` } }, [`foo`, `zod`]))
  $(
    `can have custom bullet function`,
    Tex.Tex().list({ bullet: { graphic: (i) => `(${i})` } }, [`foo`, `zod`])
  )
  $(
    `the gutter of bullets is left aligned by default`,
    Tex.Tex().list({ bullet: { graphic: (i) => String(i) } }, `abcdefghijklmnopqrstuvwxyz`.split(``))
  )
  $(
    `the gutter of bullets can be right aligned`,
    Tex.Tex().list(
      { bullet: { graphic: (i) => String(i), align: { horizontal: `right` } } },
      `abcdefghijklmnopqrstuvwxyz`.split(``)
    )
  )
})

describe(`table`, () => {
  $(
    `can have headers`,
    Tex.Tex().table(($) => $.headers([`alpha`, `bravo`]))
  )
  $(
    `can have a row`,
    Tex.Tex().table(($) => $.headers([`alpha`, `bravo`]).row(`a`, `b`))
  )
  $(
    `if a row has columns exceeding the headers they render with empty header cells`,
    Tex.Tex().table(($) => $.headers([`alpha`, `bravo`]).row(`a`, `b`, `c`))
  )
  $(
    `if headers have columns exceeding the given data they are rendered with empty column cells`,
    Tex.Tex().table(($) => $.headers([`alpha`, `bravo`, `charlie`]).row(`a`, `b`))
  )
  $(
    `data column width equals widest cell`,
    Tex.Tex().table(($) =>
      $.headers([`alpha`, `bravo`, `charlieeeeeeeeeeeeeeeeeee`, `delta`])
        .row(`a1`, `b1`, `c1`, `d1`)
        .row(`a222222222222222`, `b2`, `c2`, `d2`)
        .row(`a3`, `b333333333333333333`, `c3`, `d3`)
    )
  )
  $(
    `data column width equals widest cell (multi-line)`,
    Tex.Tex().table(($) =>
      $.headers([`alpha`, `bravo`, `charlieeeeeeeeeeeeeeeeeee`, `delta`])
        .row(`a1`, `b1`, `c1`, `d1`)
        .row(`a222222222222222\na2`, `b2`, `c2`, `d2`)
        .row(`a3`, `b333333333333333333`, `c3`, `d3`)
    )
  )
  $(
    `cell text is vertically aligned top`,
    Tex.Tex().table(($) => $.row(`alpha\napple\nankle`, `beta\nbanana`))
  )
  describe(`set`, () => {
    $(
      `custom row separator`,
      Tex.Tex().table(($) =>
        $.set({ separators: { row: ` ` } })
          .headers([`alpha`, `bravo`, `charlie`])
          .row(`a`, `b`)
      )
    )
    $(
      `custom vertical separator`,
      Tex.Tex().table(($) =>
        $.set({ separators: { column: ` ` } })
          .headers([`alpha`, `bravo`, `charlie`])
          .row(`a`, `b`)
      )
    )
  })
})

const $ = (description: string, value: Tex.RootBuilder) => {
  it(description, () => {
    expect(value.render()).toMatchSnapshot()
  })
}

$.only = (description: string, value: Tex.RootBuilder) => {
  it.only(description, () => {
    expect(value.render()).toMatchSnapshot()
  })
}
