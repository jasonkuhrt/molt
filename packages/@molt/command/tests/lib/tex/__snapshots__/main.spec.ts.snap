// Vitest Snapshot v1

exports[`block > border > bottom 1`] = `
"foo
---"
`;

exports[`block > border > left 1`] = `"|foo"`;

exports[`block > border > right 1`] = `
"foo|                   
alpha                  |
bravo bravo            |
charlie charlie charlie|"
`;

exports[`block > border > top 1`] = `
"---
foo"
`;

exports[`block > builder > can render a line 1`] = `"foo"`;

exports[`block > builder > can render nothing via null 1`] = `""`;

exports[`block > can render a line 1`] = `"foo"`;

exports[`block > can render nothing via null 1`] = `""`;

exports[`block > can render two lines 1`] = `
"foo
bar"
`;

exports[`block > padding > bottom 1`] = `
"foo

"
`;

exports[`block > padding > left 1`] = `"  foo"`;

exports[`block > padding > left multi line 1`] = `
"  foo
  bar
  qux"
`;

exports[`block > padding > top 1`] = `
"

foo"
`;

exports[`block > width > % > 100% 1`] = `
"foo bar
----------------------------------------------------------------------------------------------------
foo                                                                                                 "
`;

exports[`list > can have custom bullet 1`] = `
"- foo
- zod"
`;

exports[`list > can have custom bullet function 1`] = `
"(0) foo
(1) zod"
`;

exports[`list > can render items 1`] = `
"* foo
* bar"
`;

exports[`list > can render multi-line items 1`] = `
"* foo
* bar
  baz
  qux
* zod"
`;

exports[`list > the gutter of bullets can be right aligned 1`] = `
" 0 a
 1 b
 2 c
 3 d
 4 e
 5 f
 6 g
 7 h
 8 i
 9 j
10 k
11 l
12 m
13 n
14 o
15 p
16 q
17 r
18 s
19 t
20 u
21 v
22 w
23 x
24 y
25 z"
`;

exports[`list > the gutter of bullets is left aligned by default 1`] = `
"0  a
1  b
2  c
3  d
4  e
5  f
6  g
7  h
8  i
9  j
10 k
11 l
12 m
13 n
14 o
15 p
16 q
17 r
18 s
19 t
20 u
21 v
22 w
23 x
24 y
25 z"
`;

exports[`table > can have a row 1`] = `
"alpha | bravo
-------------
a     | b    "
`;

exports[`table > can have headers 1`] = `"alpha | bravo"`;

exports[`table > cell text is vertically aligned top 1`] = `
"alpha | beta  
apple | banana
ankle |       "
`;

exports[`table > data column width equals widest cell (multi-line) 1`] = `
"alpha            | bravo               | charlieeeeeeeeeeeeeeeeeee | delta
--------------------------------------------------------------------------
a1               | b1                  | c1                        | d1   
--------------------------------------------------------------------------
a222222222222222 | b2                  | c2                        | d2   
a2               |                     |                           |      
--------------------------------------------------------------------------
a3               | b333333333333333333 | c3                        | d3   "
`;

exports[`table > data column width equals widest cell 1`] = `
"alpha            | bravo               | charlieeeeeeeeeeeeeeeeeee | delta
--------------------------------------------------------------------------
a1               | b1                  | c1                        | d1   
--------------------------------------------------------------------------
a222222222222222 | b2                  | c2                        | d2   
--------------------------------------------------------------------------
a3               | b333333333333333333 | c3                        | d3   "
`;

exports[`table > if a row has columns exceeding the headers they render with empty header cells 1`] = `
"alpha | bravo
-----------------
a     | b     | c"
`;

exports[`table > if headers have columns exceeding the given data they are rendered with empty column cells 1`] = `
"alpha | bravo | charlie
-----------------------
a     | b    "
`;

exports[`table > set > custom row separator 1`] = `
"alpha | bravo | charlie
                       
a     | b    "
`;

exports[`table > set > custom vertical separator 1`] = `
"alpha bravo charlie
-------------------
a     b    "
`;

exports[`text > can render text 1`] = `"foo"`;
